src/unsafe-keyword.md

```diff
@@ -27,9 +27,9 @@ this can be changed by enabling the [`unsafe_op_in_unsafe_fn`] lint.
 By putting operations into an unsafe block, the programmer states that they have taken care of satisfying the extra safety conditions of all operations inside that block.
 
 Unsafe blocks are the logical dual to unsafe functions:
-where unsafe functions define a proof obligation that callers must uphold, unsafe blocks state that all relevant proof obligations have been discharged.
+where unsafe functions define a proof obligation that callers must uphold, unsafe blocks state that all relevant proof obligations of functions or operations called inside the block have been discharged.
 There are many ways to discharge proof obligations;
-for example, there could be run-time checks or data structure invariants that guarantee that certain properties are definitely true, or the unsafe block could be inside an `unsafe fn` and use its own proof obligations to discharge the proof obligations of its callees.
+for example, there could be run-time checks or data structure invariants that guarantee that certain properties are definitely true, or the unsafe block could be inside an `unsafe fn`, in which case the block can discharge the proof obligations of any unsafe functions called inside the block.
```

Comment:
Ok, so now when I read it now after I integrated those changes, it's much more clear to me what proof obligations are being met with an unsafe block inside an `unsafe fn`. The only thing I still see what stands out is that it seems like those proof obligations would exist whether or not the unsafe block was inside an unsafe function, so what really oriented my thinking was the prefix phrase `or the unsafe block could be inside an unsafe fn...`, which seemed to indicate that when the unsafe block is inside an unsafe fn that it has a special role in discharging certain proof obligations, and that's why my mind went to "oh maybe it means that the unsafe block can discharge some or all of the proof obligations of the `unsafe fn` it is residing in". I think I understand that to NOT be the case, so I'm wondering if the prefix phrase should be removed so it would say:

```suggestion
for example, there could be run-time checks or data structure invariants that guarantee that certain properties are definitely true, or an unsafe block can discharge the proof obligations of any unsafe functions called inside the block.
```

Reply:
Or maybe, in other words: does the requirement that an unsafe block discharge any proof obligations of unsafe functions that it calls change if the block exists within a safe or unsafe function?

Reply:
No, that requirement doesn't change.

If you put the words "use the proof obligations of that function to" back then I think the reference to the parent unsafe function will make more sense.

Reply:
>  so what really oriented my thinking was the prefix phrase or the unsafe block could be inside an unsafe fn..., which seemed to indicate that when the unsafe block is inside an unsafe fn that it has a special role in discharging certain proof obligations

The point was that inside an unsafe fn, there are extra preconditions we can use. When the unsafe block is in a regular safe fn, this is not the case -- the callee's preconditions have to be discharged based on whatever the function does itself, and private invariants, but it cannot "propagate" the requirements further outwards.

> an unsafe block can discharge the proof obligations of any unsafe functions called inside the block

It *must* discharge those proof obligations! The question is where the information to discharge the obligations comes from:
- local checks
- data structure invariants
- (only possible if this is inside an `unsafe fn`) preconditions of the current function

Reply:
You removed the crucial part from @mattheww's proposal! "use the proof obligations of that function" was load-bearing, after removing it this sentence now doesn't make a ton of sense any more.

Reply:
Ok, I think I understand better now. It's using the proof obligations of the `unsafe fn` to discharge the proof obligations of other unsafe functions called in an unsafe block within the original `unsafe fn`, which wouldn't make any sense if the original function was safe, because then it would have no proof obligations. I understand why that part needs to be there now. I've added that phrase back in.